lib/cdo/redshift_import.rb

```diff
@@ -0,0 +1,56 @@
+require 'cdo/redshift'
+
+class RedshiftImport
+  # Database Migration Service Replication Tasks load data from Aurora into staging Redshift tables with a prefix.
+  TEMP_TABLE_PREFIX = '_import_'.freeze
+
+  # DMS Replication Tasks import data into temporary / staging tables prefixed with "_import_".
+  # Complete import of staged data from production database into Redshift by truncating each
+  # target table, moving all of the rows from each import table to its corresponding target table, and then dropping
+  # each temporary / staging import table.
+  # @param schemas [String] Array of Redshift schemas in which to complete data import.
+  def self.complete_table_import(schemas)
+    schemas.each do |schema|
+      temporary_import_tables(schema).each do |import_table|
+        target_table = import_table.partition(TEMP_TABLE_PREFIX).last
+        truncate_table(schema, target_table)
+        move_rows(schema, import_table, schema, target_table)
+        drop_table(schema, import_table)
+      end
+    end
+  end
+
+  # Get names of tables in a Redshift schema that were loaded via DMS to stage imported data from the database.
+  # @param schema [String] The Redshift schema to search for database import staging tables.
+  # Returns Array of Redshift table names.
+  def self.temporary_import_tables(schema)
+    redshift_client = RedshiftClient.instance
+    query = <<~SQL
+      SET search_path TO #{schema};
+      SELECT DISTINCT t.tablename
+      FROM pg_table_def t LEFT JOIN pg_indexes i ON t.tablename = i.indexname
+      WHERE  i.indexname IS NULL -- Don't count primary keys, which appear in pg_table_def, as a table.
+        AND  t.schemaname ='#{schema}'
+        AND t.tablename LIKE '#{TEMP_TABLE_PREFIX}%';
+    SQL
+    redshift_client.exec(query).map {|row| row['tablename']}
+  end
+
+  def self.truncate_table(schema, table)
+    redshift_client = RedshiftClient.instance
+    redshift_client.exec "TRUNCATE #{schema}.#{table};"
+  end
+
+  def self.drop_table(schema, table)
+    redshift_client = RedshiftClient.instance
+    redshift_client.exec "DROP TABLE IF EXISTS #{schema}.#{table};"
+  end
+
+  def self.move_rows(source_schema, source_table, target_schema, target_table)
+    redshift_client = RedshiftClient.instance
+    # IGNOREEXTRA ensures that moving rows to the target table does not fail if a column was added to the source
+    # MySQL table.
+    # TODO: (suresh) Switch to a technical design that supports schema changes in MySQL propagating to Redshift.
+    redshift_client.exec "ALTER TABLE #{target_schema}.#{target_table} APPEND FROM #{source_schema}.#{source_table} IGNOREEXTRA;"
```

Comment:
@bencodeorg The `ALTER TABLE ... APPEND` approach does not permit schema changes in the source MySQL table to propagate automatically to the Redshift table (I assume DMS Change Data Capture supports this?).  One table fails during the execution of this method during the weekly import this morning because it had a new column.  `IGNOREEXTRA` is a temporary workaround to get the data moving smoothly, but I need some assist finding an alternate solution.

Could we just drop the target table and rename the staging table?  How would that impact permissions?

Reply:
I'm pretty sure tables created in a given schema inherit permissions from the schema (otherwise we would have had permissions issues in the past when importing new production tables to Redshift via DMS).

So, I don't know if you would be able to just rename the staging table, or if you'd have to create a new table with the same schema as the staging table? I'm imagining you add a step that checks each table before the append that checks that the tables are the same, something [like this](https://dataedo.com/kb/query/postgresql/compare-tables-and-columns-in-two-schemas)?

Reply:
> I'm pretty sure tables created in a given schema inherit permissions from the schema (otherwise we would have had permissions issues in the past when importing new production tables to Redshift via DMS).

That reasoning makes sense.  Our Full Load (no-CDC) Replication Tasks currently `DROP` the target table and then re-create them from scratch, and as you pointed out, we sometimes add new tables to the Replication Tasks rules.

> So, I don't know if you would be able to just rename the staging table, or if you'd have to create a new table with the same schema as the staging table? I'm imagining you add a step that checks each table before the append that checks that the tables are the same, something [like this](https://dataedo.com/kb/query/postgresql/compare-tables-and-columns-in-two-schemas)?

I think our new approach would be - 

for each staging table (`_import_`):
1. drop the target table
2. rename the staging table to drop the `_import_` prefix

Reply:
That sounds reasonable -- I thought this whole truncate and append approach was related to the fact that we decided we couldn't rename things? Orrr, writing this out, was it that we couldn't move between schemas, and in this case your `_import_` tables are already in the same schema so that would work?

Reply:
I think when you and I prototyped staging tables in October/November 2019, we imported the staging table into a staging schema (`dashboard_production_temp`), and then attempted to move/rename the staging table into the target schema.  I think those tests found that permissions were not set correctly and you found `ALTER TABLE ... APPEND` as an alternative.